File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Golang/ContaBancaria.go
package main

import (
	"errors"
	"fmt"
)

type ContaBancaria struct {
	titular string
	saldo   float64
}

func (c *ContaBancaria) Depositar(valor float64) {
	if valor > 0 {
		c.saldo += valor
		fmt.Printf("Depósito de R$%.2f realizado com sucesso! Novo saldo: R$%.2f\n", valor, c.saldo)
	} else {
		fmt.Println("O valor de depósito deve ser positivo!")
	}
}

func (c *ContaBancaria) Sacar(valor float64) error {
	if valor > 0 {
		if valor <= c.saldo {
			c.saldo -= valor
			fmt.Printf("Saque de R$%.2f realizado com sucesso! Novo saldo: R$%.2f\n", valor, c.saldo)
			return nil
		}
		return errors.New("Saldo insuficiente!")
	}
	return errors.New("o valor de saque deve ser positivo")
}

func (c *ContaBancaria) ObterSaldo() float64 {
	fmt.Printf("Saldo de R$%.2f\n", c.saldo)
	return c.saldo
}

File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Python/ContaBancaria.py
class SaldoInsuficiente(Exception):
    def __init__(self, saldo, valor):
        super().__init__(f"ERROR: Saldo insuficiente!\nSaque de R$ {valor} falhou. Saldo disponível: R$ {saldo}")
        self.saldo = saldo
        self.valor = valor

class ContaBancaria:
    def __init__(self, titular, saldo):
        self.titular = titular
        self.saldo = saldo

    def depositar(self, valor):
        self.saldo += valor
        print(f"Deposito de R$ {valor} feito, Saldo disponível: R$ {self.saldo}")
    
    def sacar(self, valor):
        if (valor > self.saldo):
            raise SaldoInsuficiente(self.saldo, valor)
        else:
            self.saldo -= valor
            print(f"Saque de R$ {valor} feito, Saldo disponível: R$ {self.saldo}")

    def obter_saldo(self):
        print(f"Saldo disponível: R$ {self.saldo}")
        return self.saldo
            
